Backend/Middleware-APD/src/main/java/middlewareapd/service/MiddlewareService.java

```diff
@@ -0,0 +1,116 @@
+package middlewareapd.service;
+
+import java.time.LocalDateTime;
+import java.util.Map;
+import java.util.concurrent.locks.ReadWriteLock;
+
+import middlewareapd.exception.UnauthorizedException;
+import middlewareapd.exception.UserNotFoundException;
+import middlewareapd.model.JWToken;
+import middlewareapd.repository.MockJWTRepository;
+import middlewareapd.util.LockFactory;
+import middlewareapd.util.ValidationUtil;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * MiddlewareService handles operations related to JWT validation and management.
+ * It provides methods to validate JWT tokens, update access times, and interact with the JWT repository.
+ */
+public class MiddlewareService {
+    private final MockJWTRepository jwtRepository;
+    private static final Logger logger = LoggerFactory.getLogger(MiddlewareService.class);
+
+    /**
+     * Constructs a MiddlewareService with the specified JWT repository.
+     *
+     * @param jwtRepository the JWT repository to interact with.
+     */
+    public MiddlewareService(MockJWTRepository jwtRepository) {
+        this.jwtRepository = jwtRepository;
+    }
+
+    /**
+     * Validates the given JWT token and updates the last access time for the associated user.
+     *
+     * @param token the JWT token to validate.
+     * @return a map containing the UUID and other information if validation is successful.
+     */
+    public Map<String, Object> checkJwt(String token) {
+        ReadWriteLock jwtLock = LockFactory.getRWLock("JWTValidationLock");
+        String uuid;
+        Map<String, Object> result;
+
+        try {
+            // Validate the JWT and retrieve the result outside of the lock
+            result = ValidationUtil.validateJwt(token, jwtRepository);
+            uuid = (String) result.get("uuid");
+
+            // Update last access time within a write lock
+            @SuppressWarnings("unused")
+            JWToken dbJwt = updateLastAccess(uuid, jwtLock);
+            logger.info("Successfully validated JWT for UUID: {}", uuid);
+            return result;
+        } catch (UnauthorizedException e) {
+            handleUnauthorizedException(e);
+            throw e; // Return an empty map or handle accordingly
+        } catch (UserNotFoundException e) {
+            handleUserNotFoundException(e);
+            throw e; // Return an empty map or handle accordingly
+        } catch (Exception e) {
+            handleUnexpectedException(e);
+            throw e; // Return an empty map or handle accordingly
+        }
+    }
+
+    /**
+     * Updates the last access time for the given UUID and saves the updated token.
+     *
+     * @param uuid the UUID of the user whose access time is to be updated.
+     * @return the updated JWToken, or null if the token was not found.
+     */
+    private JWToken updateLastAccess(String uuid, ReadWriteLock jwtLock) {
+        JWToken dbJwt = jwtRepository.getTokenByUuid(uuid); // Read without locking
```

Comment:
retrievedToken?

Reply:
Modified function to return null also since result isn't used